source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,410 @@
+    }
+  } else {
+    // The only valid pseudo header for responses is :status. If the header name starts with ":"
+    // and it's not ":status", then the header name is an unknown pseudo header.
+    return {HeaderEntryValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidPseudoHeader};
+  }
+
+  // Validate the header value
+  return validateGenericHeaderValue(value);
+}
+
+::Envoy::Http::HeaderValidator::RequestHeaderMapValidationResult
+Http2HeaderValidator::validateRequestHeaderMap(::Envoy::Http::RequestHeaderMap& header_map) {
+  static const absl::node_hash_set<absl::string_view> kAllowedPseudoHeadersForConnect = {
+      ":method",
+      ":authority",
+  };
+
+  static const absl::node_hash_set<absl::string_view> kAllowedPseudoHeaders = {
+      ":method", ":scheme", ":authority", ":path"};
+
+  absl::string_view path = header_map.getPathValue();
+
+  // Step 1: verify that required pseudo headers are present.
+  //
+  // The method pseudo header is always mandatory.
+  if (header_map.getMethodValue().empty()) {
+    return {RequestHeaderMapValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidMethod};
+  }
+
+  auto is_connect_method = header_map.method() == header_values_.MethodValues.Connect;
+  auto is_options_method = header_map.method() == header_values_.MethodValues.Options;
+  bool path_is_asterisk = path == "*";
+  bool path_is_absolute = !path.empty() && path.at(0) == '/';
+
+  if (!is_connect_method && (header_map.getSchemeValue().empty() || path.empty())) {
+    // If this is not a connect request, then we also need the scheme and path pseudo headers.
+    // This is based on RFC 9113, https://www.rfc-editor.org/rfc/rfc9113#section-8.3.1:
+    //
+    // All HTTP/2 requests MUST include exactly one valid value for the ":method", ":scheme", and
+    // ":path" pseudo-header fields, unless they are CONNECT requests (Section 8.5). An HTTP
+    // request that omits mandatory pseudo-header fields is malformed (Section 8.1.1).
+    auto details = path.empty() ? UhvResponseCodeDetail::get().InvalidUrl
+                                : UhvResponseCodeDetail::get().InvalidScheme;
+    return {RequestHeaderMapValidationResult::Action::Reject, details};
+  } else if (is_connect_method) {
+    // If this is a CONNECT request, :path and :scheme must be empty and :authority must be
+    // provided. This is based on RFC 9113,
+    // https://www.rfc-editor.org/rfc/rfc9113#section-8.5:
+    //
+    //  * The ":method" pseudo-header field is set to CONNECT.
+    //  * The ":scheme" and ":path" pseudo-header fields MUST be omitted.
+    //  * The ":authority" pseudo-header field contains the host and port to connect to (equivalent
+    //    to the authority-form of the request-target of CONNECT requests; see Section 3.2.3 of
+    //    [HTTP/1.1]).
+    //
+    // TODO - Add support for extended CONNECT requests: envoy.reloadable_features.use_rfc_connect
```

Comment:
wasn't that draft abandoned?  I think the reloadable flag removes extended connect support so I think you can remove this TODO